crates/chain/src/keychain/txout_index.rs

```diff
@@ -11,6 +11,75 @@ use core::{fmt::Debug, ops::Deref};
 
 use crate::Append;
 
+/// Represents updates to the derivation index of a [`KeychainTxOutIndex`].
+/// It maps each keychain `K` to a descriptor and its last revealed index.
+///
+/// It can be applied to [`KeychainTxOutIndex`] with [`apply_changeset`]. [`ChangeSet] are
+/// monotone in that they will never decrease the revealed derivation index.
+///
+/// [`KeychainTxOutIndex`]: crate::keychain::KeychainTxOutIndex
+/// [`apply_changeset`]: crate::keychain::KeychainTxOutIndex::apply_changeset
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Deserialize, serde::Serialize),
+    serde(
+        crate = "serde_crate",
+        bound(
+            deserialize = "K: Ord + serde::Deserialize<'de>",
+            serialize = "K: Ord + serde::Serialize"
+        )
+    )
+)]
+#[must_use]
+pub struct ChangeSet<K>(pub BTreeMap<K, (Descriptor<DescriptorPublicKey>, u32)>);
```

Comment:
The problem with this design, is that every time we need to increase the last derivation index, we need to include the descriptor in the changeset. This is rather expensive.

What do you think about having two fields? One for descriptors, one for last derivation indices?

Reply:
Could it be like this:

```
ChangeSet<K> {
    /// Descriptors indexed by keychain. Note that in the case a changeset is applied to an [`KeychainTxOutIndex`] that already has a descriptor under that keychain then it will choose the greatest `Descriptor` according to `Ord`. 
    pub descriptors: BTreeMap<K, Descriptor>,
    /// Index of the script pubkey that was last revealed
    pub last_revealed: BTreeMap<Descriptor, u32>,
}
```
This links descriptors to last derivation indices rather than keychain. We could do this inside the `KeychainTxOutIndex` as well.

A more broad question is can we take moment to redesign this without `K` at all? We could use the `(Descriptor, u32)` as the key in the underlying `Spk` tracker. This would be wasteful from a memory perspective so we could compress this down to `32` bytes with sha256 -- but it's still a bit wasteful. Other option is to come up with a internal index for each descriptor as they are added.

I think I'm against this idea but thought I should bring it up. Offloading the problem of labeling Descriptors to the user keeps things efficient internally and I think you almost always want to label descriptors with some semantic meaning inside your application.

Reply:
> This links descriptors to last derivation indices rather than keychain.

@LLFourn Is this so that we can avoid setting the last revealed index to a keychain with no descriptor?

Reply:
It's to avoid a changeset changing the descriptor of a keychain but inheriting the last seen of the old descriptor. Thinking further doing literally what I suggested above is a terrible idea because descriptors can be arbitrarily large (you already pointed this out). Not sure if necessary -- I'd very happy if you and/or @danielabrozzoni to make the call about the design here. I think my preference has changed to what you suggested in the OP.

Reply:
> I think I'm against this idea but thought I should bring it up. Offloading the problem of labeling Descriptors to the user keeps things efficient internally and I think you almost always want to label descriptors with some semantic meaning inside your application.

I agree, I think it's useful from the user perspective to have a way to label descriptor, and so I'd prefer not to remove it.